web/src/types/option.ts

```diff
@@ -0,0 +1,199 @@
+import { ToString } from './utils';
+import { Err, Ok, Result } from './result';
+
+export const Some = <S extends ToString>(value: S) => {
+  return new Option<S>(false, value);
+};
+
+export const None = () => {
+  return new Option<never>(true);
+};
+
+export class Option<S extends ToString> {
+  private is_none: boolean;
+  private value: S | never | undefined;
+
+  constructor(is_none: boolean, value?: S) {
+    this.is_none = is_none;
+    this.value = value;
+  }
+
+  isSome(): this is Option<S> {
+    return !this.is_none;
+  }
+
+  isSomeAnd(op: (value: Readonly<S>) => boolean): this is Option<S> {
+    return this.isSome() && op(this.value as S);
+  }
+
+  isNone(): this is Option<never> {
+    return this.is_none;
+  }
+
+  match<T, K>(matcher: {
+    some: (value: Readonly<S>) => T;
+    none: () => K;
+  }): T | K {
+    // eslint-disable-next-line unicorn/no-array-callback-reference
+    return this.is_none ? matcher.none() : matcher.some(this.value as S);
+  }
+
+  map<T extends ToString>(fn: (value: Readonly<S>) => T): Option<T> {
+    return this.match({
+      some: (value) => Some(fn(value)),
+      none: None,
+    });
+  }
+
+  mapOr<T>(defaultValue: T, fn: (value: Readonly<S>) => T): T {
+    return this.match({
+      some: (value) => fn(value),
+      none: () => defaultValue,
+    });
+  }
+
+  mapOrElse<T>(defaultFn: () => T, fn: (value: Readonly<S>) => T): T {
+    return this.match({
+      some: (value) => fn(value),
+      none: defaultFn,
+    });
+  }
+
+  okOr<E extends ToString>(error: E): Result<S, E> {
+    return this.match({
+      some: (value) => Ok(value),
+      none: () => Err(error),
+    });
+  }
+
+  okOrElse<E extends ToString>(err: () => E): Result<S, E> {
+    return this.match({
+      some: (value) => Ok(value),
+      none: () => Err(err()),
+    });
+  }
+
+  and<T extends ToString>(other: Option<T>): Option<T> {
+    return this.match({
+      some: () => other,
+      none: None,
+    });
+  }
+
+  andThen<T extends ToString>(
+    fn: (value: Readonly<S>) => Option<T>,
+  ): Option<T> {
+    return this.match({
+      some: (value) => fn(value),
+      none: None,
+    });
+  }
+
+  filter(predicate: (value: Readonly<S>) => boolean): Option<S> {
+    return this.match({
+      some: (value) => (predicate(value) ? Some(value) : None()),
+      none: None,
+    });
+  }
+
+  or<T extends ToString>(other: Option<S | T>): Option<S | T> {
+    return this.match({
+      some: (value) => Some(value),
+      none: () => other,
+    });
+  }
+
+  orElse<T extends ToString>(fn: () => Option<S | T>): Option<S | T> {
+    return this.match({
+      some: (value) => Some(value),
+      none: fn,
+    });
+  }
+
+  xor<T extends ToString>(other: Option<S | T>): Option<S | T> {
+    if (this.isSome() && other.isNone()) {
+      return this;
+    } else if (this.isNone() && other.isSome()) {
+      return other;
+    } else {
+      return None();
+    }
+  }
+
+  take(): Option<S> {
```

Comment:
`take()` isn't really useful in a language without borrow checking IMO, and it's kind of surprising for a type that I'd otherwise expect to be immutable.

Reply:
Agreed, removed in bad24a5